test/extended/networking/ipsec.go

```diff
@@ -0,0 +1,676 @@
+			config.dstNodeConfig.hostInf, err = findBridgePhysicalInterface(oc, node2.Name, "br-ex")
+			o.Expect(err).NotTo(o.HaveOccurred())
+			for _, address := range node2.Status.Addresses {
+				if address.Type == corev1.NodeInternalIP {
+					config.dstNodeConfig.nodeIP = address.Address
+					break
+				}
+			}
+			o.Expect(config.dstNodeConfig.nodeIP).NotTo(o.BeEmpty())
+		})
+
+		g.AfterEach(func() {
+			// Restore the cluster back into original state after running all the tests.
+			g.By("restoring ipsec config into original state")
+			err := configureIPsecMode(oc, config.ipsecMode)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			waitForIPsecConfigToComplete(oc, config.ipsecMode)
+
+			g.By("remove right node ipsec configuration")
+			err = oc.AsAdmin().Run("delete").Args("-f", rightNodeIPsecConfigYaml).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("remove left node ipsec configuration")
+			err = oc.AsAdmin().Run("delete").Args("-f", leftNodeIPsecConfigYaml).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("undeploy nmstate handler")
+			err = undeployNmstateHandler(oc)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("removing IPsec certs from worker nodes")
+			err = deleteNSCertMachineConfig(oc)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			o.Eventually(func() bool {
+				ready, err := isWorkerMachineConfigPoolsReady(oc, nsCertMachineConfigName, false)
+				o.Expect(err).NotTo(o.HaveOccurred())
+				return ready
+			}, 20*time.Minute, 180*time.Second).Should(o.BeTrue())
+		})
+
+		g.DescribeTable("check traffic between local pod to a remote pod [apigroup:config.openshift.io] [Suite:openshift/network/ipsec]", func(mode v1.IPsecMode) {
+			o.Expect(config).NotTo(o.BeNil())
+			g.By("validate pod traffic before changing IPsec configuration")
+			// Ensure pod traffic is working with right encapsulation before rolling out IPsec configuration.
+			checkTraffic(config.ipsecMode)
+			g.By(fmt.Sprintf("configure IPsec in %s mode and validate pod traffic", mode))
+			err := configureIPsecMode(oc, mode)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			waitForIPsecConfigToComplete(oc, mode)
+			// Ensure pod traffic is working with right encapsulation after rolling out IPsec configuration.
+			checkTraffic(mode)
+		},
+			g.Entry("with IPsec in full mode", v1.IPsecModeFull),
+			g.Entry("with IPsec in external mode", v1.IPsecModeExternal),
+			g.Entry("with IPsec in disabled mode", v1.IPsecModeDisabled),
+		)
+
+		// This test sets up IPsec mode to External, chooses two worker nodes as left and right
+		// servers for IPsec configuration, generate and import certs into their nss db, configure
+		// its ipsec conf via nmstate and then check node traffic between them are IPsec encrypted.
+		// The NS IPsec is being tested only with External IPsec mode because when Full mode is
+		// enabled, then it already creates IPsec mesh between nodes for the pod traffic, so NNCP
+		// doesn't have any effect on the node.
```

Comment:
Ok @jcaamano , Nice suggestion. so we may need another test to cover this and make use of tcpdump pod itself for sending ping traffic to cover host networked pods. will do with separate `DescribeTable`.

Reply:
I think ideally you would have both pod network and host network traffic going and check:

E/W enabled: check for presence of esp traffic, check for absence of unencrypted geneve traffic, check for presence of normal unencrypted traffic
N/S enabled or N/S & E/W enabled: check for esp traffic, check for absence of unencrypted traffic

Reply:
Oh I guess the caveat here is that for Full N/S & E/W enabled, we can't tell if the pod network traffic is encrypted due to N/S or E/W ipsec. I don't even know if configuring both as you are doing would work.

Reply:
We should additionally check for ICMP traffic for host networked pod, that would be always unencrypted for IPsec EW, when NS is enabled between two nodes, this would change to encrypted traffic. Hope this would work. WDYT ?

Reply:
done, please take a look now.